### PR TITLE
DOC: More on pandas comparison

### PR DESCRIPTION
**Pull request by max-sixty**
Follow up from the mailing list:
- Added some more thoughts on the multi-dimensional comparison. Some of this is opinionated (in concepts not arguments) so I'd appreciate feedback on both the concepts and language. 
- Removed some of the specific comparison with `NDPanel` etc, given those are removed
- Placeholder for something on the Explicit API (wanted to get this version out, consistent with me leaving less work hanging)

### CODE REVIEW

**Comment by max-sixty** [comment]
We could also add some references from people who've written about similar ideas; e.g.
- http://nlp.seas.harvard.edu/NamedTensor
- https://twitter.com/TedPetrou has written about the pandas API
- a bit less relevant, but https://github.com/InvestmentSystems/static-frame is the issue I referenced on the mailing list

I'll also find the "I tried to document the pandas indexing API" issue

File: doc/faq.rst [hunk 1]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
+    than their position. By allowing us to use their names, our code becomes 
+    easier to write and much easier to read; 
+    e.g. ``da.isel(time=0)`` rather than ``da[:, :, 0]``.
+
+  - *Broadcasting*: Xarray can broadcast operations to an array, meaning it
+    applies the operation similarly regardless of the number of dimensions. 
+    This allows our code to be more abstract, requiring far fewer special cases. 
+
+    For example, ``da - 1`` would make every pixel slightly darker, whether 
+    we're dealing with a single or multiple frames, color or B&W, etc.
```
Comment (shoyer) [inline doc/faq.rst]:
```suggestion
    we're dealing with a single or multiple frames, color or grayscale, etc. 
```

File: doc/faq.rst [hunk 2]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
+    e.g. ``da.isel(time=0)`` rather than ``da[:, :, 0]``.
+
+  - *Broadcasting*: Xarray can broadcast operations to an array, meaning it
+    applies the operation similarly regardless of the number of dimensions. 
+    This allows our code to be more abstract, requiring far fewer special cases. 
+
+    For example, ``da - 1`` would make every pixel slightly darker, whether 
+    we're dealing with a single or multiple frames, color or B&W, etc. 
+    
+    Similarly ``da - da.mean('row', 'column')`` would normalize the values
```
Comment (shoyer) [inline doc/faq.rst]:
```suggestion
    Similarly ``da - da.mean(dim=['row', 'column'])`` would normalize the values 
```

File: doc/faq.rst [hunk 3]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
+  - *Broadcasting*: Xarray can broadcast operations to an array, meaning it
+    applies the operation similarly regardless of the number of dimensions. 
+    This allows our code to be more abstract, requiring far fewer special cases. 
+
+    For example, ``da - 1`` would make every pixel slightly darker, whether 
+    we're dealing with a single or multiple frames, color or B&W, etc. 
+    
+    Similarly ``da - da.mean('row', 'column')`` would normalize the values 
+    based on the mean value in its width and height, again regardless of whether
+    we have a single or multiple frames, color or B&W, etc.
```
Comment (shoyer) [inline doc/faq.rst]:
```suggestion
    we have a single or multiple frames, color or grayscale, etc.
```

File: doc/faq.rst [hunk 4]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
+
+  - *Alignment*: Xarray can package arrays which share some, but not all, of
+    their dimensions in a dataset and align the dimensions they do share. 
+    For example, if we calculated the brightness of our movie images
+    using ``da.mean('row', 'column', 'color')`, we'd have an array with a time
```
Comment (shoyer) [inline doc/faq.rst]:
```suggestion
    using ``da.mean(dim=['row', 'column', 'color'])``, we'd have an array with a time
```

File: doc/faq.rst [hunk 5]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
+~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes we want to use higher dimensional arrays, or arrays for 
+which the names of dimensions have some semantic significance we want to retain.
 Pandas gets a lot of things right, but many science, engineering and complex
 analytics use cases need fully multi-dimensional data structures.
 
-How do xarray data structures differ from those found in pandas?
-----------------------------------------------------------------
+For example, consider the images of a movie, represented as an array of pixels with
```
Comment (shoyer) [inline doc/faq.rst]: I'm not sure the movie example is the most compelling comparison to pandas. I think it's pretty obvious that a DataFrame is a poor fit for a movie -- the obvious data structure to use would be a NumPy array.

Maybe something 3D that could fit into a Panel or multi-indexed DataFrame would be a better choice? Examples that come to mind:
- open/high/low/close stock data (2D array of shape time/company in an xarray.Dataset) -- this was the textbook example for `pandas.Panel`
- [inference data in ArviZ](https://arviz-devs.github.io/arviz/notebooks/InferenceDataCookbook.html) with dimensions like (chain, draw, ...)

Comment (max-sixty) [reply doc/faq.rst]: Yes, I agree, the movie example isn't great as a comparison to pandas (good for numpy)
Is there something that's more generally intuitive than finance or Bayesian stats? I think weather might be our best bet! Open to others

File: doc/faq.rst [hunk 6]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
-  'major_axis', etc.) but the dimension names refer to order, not their
-  meaning. You can't specify an operation as to be applied along the "time"
-  axis.
-- You often have to manually convert collections of pandas arrays
-  (Series, DataFrames, etc) to have the same number of dimensions.
-  In contrast, this sort of data structure fits very naturally in an
-  xarray ``Dataset``.
-
-You can :ref:`read about switching from Panels to Xarray here <panel transition>`.
+if data can be sensibly described as "rows and columns", pandas is probably the
```
Comment (dcherian) [inline doc/faq.rst]:
```suggestion
if the data can be sensibly described as "rows and columns", pandas is probably the
```

File: doc/faq.rst [hunk 7]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
+  - *Alignment*: Xarray can package arrays which share some, but not all, of
+    their dimensions in a dataset and align the dimensions they do share. 
+    For example, if we calculated the brightness of our movie images
+    using ``da.mean('row', 'column', 'color')`, we'd have an array with a time
+    dimension. When we add that to a dataset and select the first point in time,
+    ``ds.isel(time=0)``, we'll receive both the frame and the first brightness
+    value. 
+
+    Even following best-pracitices around tidy tabular data, this approach is
+    very difficult to replicate using tabular data.
```
Comment (dcherian) [inline doc/faq.rst]:
```suggestion
    very difficult or at least unintuitive to replicate using tabular data.
```

**Comment by max-sixty** [comment]
Thanks @dcherian ! First time I've used that feature. I'll aim to finish this off shortly

File: doc/faq.rst [hunk 8]
```diff
@@ -15,45 +15,61 @@ Why is pandas not enough?
...
-  axis.
-- You often have to manually convert collections of pandas arrays
-  (Series, DataFrames, etc) to have the same number of dimensions.
-  In contrast, this sort of data structure fits very naturally in an
-  xarray ``Dataset``.
-
-You can :ref:`read about switching from Panels to Xarray here <panel transition>`.
+if the data can be sensibly described as "rows and columns", pandas is probably the
+right choice.  However, xarray has some distinct advantages: higher dimensionality
+and an explicit, production-ready API.
```
Comment (mathause) [inline doc/faq.rst]: this sounds a bit as if the pandas API was not production-ready

Comment (max-sixty) [reply doc/faq.rst]: I'll soften to a comparative

Comment (shoyer) [reply doc/faq.rst]: I'm not sure either xarray or pandas are production ready :)

**Pull request closed without merging**